sled-agent/src/instance.rs

```diff
@@ -492,32 +525,147 @@ impl Instance {
             state: InstanceStates::new(initial.runtime),
             running_state: None,
             lazy_nexus_client,
+            instance_ticket: ticket,
         };
 
         let inner = Arc::new(Mutex::new(instance));
 
         Ok(Instance { inner })
     }
 
+    pub async fn current_state(&self) -> InstanceRuntimeState {
+        let inner = self.inner.lock().await;
+        inner.state.current().clone()
+    }
+
+    /// Ensures that a Propolis process exists for this instance, then sends it
+    /// an instance ensure request.
+    async fn propolis_ensure(
+        &self,
+        inner: &mut MutexGuard<'_, InstanceInner>,
+        migration_params: Option<InstanceMigrationTargetParams>,
+    ) -> Result<(), Error> {
+        if let Some(running_state) = inner.running_state.as_ref() {
+            inner
+                .propolis_ensure(
+                    &running_state.client,
+                    &running_state.running_zone,
+                    migration_params,
+                )
+                .await?;
+        } else {
+            // If there's no Propolis yet, and this instance is not being
+            // initialized via migration, immediately send a state update to
+            // Nexus to reflect that the instance is starting (so that the
+            // external API will display this state while the zone is being
+            // started).
+            //
+            // Migration targets don't do this because the instance is still
+            // logically running (on the source) while the target Propolis is
+            // being launched.
+            if migration_params.is_none() {
+                inner.state.transition(InstanceState::Starting);
+                inner.publish_state_to_nexus().await?;
```

Comment:
If this fails, are we relying on the saga to unwind our state?

If we exit with an error here, the `InstanceState` will be `Starting`, even though no propolis has actually been started. Perhaps we should handle an error here by transitioning to `Failed`?

Reply:
I see we're doing something similar with `setup_result` below - maybe these transitions should be lumped into that block?

Reply:
Good catch. Fixed in 7672fcfa5.